ouroboros-consensus/test-consensus/Test/Dynamic/Util/NodeTopology.hs

```diff
@@ -0,0 +1,129 @@
+  , genNodeTopology
+  , shrinkNodeTopology
+  , meshNodeTopology
+  , bottleneckSizeNodeTopology
+  ) where
+
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import           GHC.Stack (HasCallStack)
+import           Test.QuickCheck
+
+import           Ouroboros.Consensus.Node.ProtocolInfo.Abstract
+import           Ouroboros.Consensus.NodeId
+import           Ouroboros.Consensus.Util.Condense
+import           Ouroboros.Consensus.Util.Orphans ()
+
+{-------------------------------------------------------------------------------
+  Node Topologies
+-------------------------------------------------------------------------------}
+
+-- | Which /lesser/ nodes each node connects to
+--
+-- INVARIANT: for each mapping @n -> ms@, @n > m@ for each @m@.
+--
+-- INVARIANT: each mapping is non-empty, except for @CoreNodeId 0@
+--
+-- INVARIANT: each mapping is sorted strictly ascending
+--
+-- INVARIANT: there is a mapping for each @CoreNodeId@ in the test, @0 .. n -
+-- 1@
+--
+newtype NodeTopology = NodeTopology (Map CoreNodeId [CoreNodeId])
+  deriving (Eq, Show)
+
+instance Condense NodeTopology where
+  condense top@(NodeTopology m)
+    | top == mesh = "meshNodeTopology (NumCoreNodes " ++ show (Map.size m) ++ ")"
+    | otherwise = condense
+      [ (fromCoreNodeId nid, map fromCoreNodeId nids)
+      | (nid, nids) <- Map.toAscList m ]
+    where
+      mesh = meshNodeTopology (NumCoreNodes (Map.size m))
+
+-- | Connect every pair of nodes
+--
+meshNodeTopology ::
+     NumCoreNodes
+     -- ^ PRECONDITION: non-negative
+  -> NodeTopology
+meshNodeTopology numCoreNodes =
+    NodeTopology $
+    Map.fromList $
+    [ (nid, enumCoreNodes (NumCoreNodes i))
+    | nid@(CoreNodeId i) <- enumCoreNodes numCoreNodes ]
+
+-- | Generate a 'NodeTopology' consistent with the given properties
+--
+genNodeTopology ::
+     HasCallStack
+  => NumCoreNodes
+     -- ^ PRECONDITION: non-negative
+  -> Gen NodeTopology
+genNodeTopology numCoreNodes@(NumCoreNodes n)
+  | n < 0 = error $ "Unsatisfiable parameters: "
+    ++ show numCoreNodes
+
+  | otherwise = do
+    let genNeighbors me@(CoreNodeId i) = case i of
+          0 -> pure (me, [])
+          _ ->
+            fmap ((,) me) $
+            flip suchThat (not . null) $
+            sublistOf (enumCoreNodes (NumCoreNodes i))
+
+    fmap (NodeTopology . Map.fromList) $
+      mapM genNeighbors (enumCoreNodes numCoreNodes)
+
+-- | Shrink a node topology
+--
+-- The new topologies must be usable with the same number of nodes and slots as
+-- the old topology
+--
+shrinkNodeTopology :: NodeTopology -> [NodeTopology]
+shrinkNodeTopology top@(NodeTopology m)
+    | top == mesh = []
+    | otherwise = [mesh]
+  where
+    mesh = meshNodeTopology (NumCoreNodes (Map.size m))
+    -- TODO more sophisticated shrinks. I anticipate that they'll need to use
+    -- 'Test.QuickCheck.Shrinking' or else risk very slow responses
+
+-- | Partial; @error@ for a node not in the plan
```

Comment:
Particular reason to return an error rather than a `Maybe` here?

Reply:
This case is a violation of an invariant on `NodeTopology`, so this seems fair to me. In other words, I never anticipate code handling the `Nothing` case, so it'd just be pushing noise to the call sites.

The prereq can be checked at the call-site easily enough, if my anticipation is wrong.